File: py/bedrock/evidence.py
"""Offline evidence verification (spec §4.4, §8.2, §13.2).
Mirrors src/evidence.ts. Never contacts the registry; never dispatches."""
from __future__ import annotations

from .canon import canonical_string
from .crypto import D, hex_decode, ed25519_verify, b64url_decode, S
from .schema import (
    v_evidence, v_audit_entry, v_checkpoint, v_pin_update,
)
from .evaluator import evaluate
from .bundle import canonical_equal, verify_bundle

ZERO_HASH = "0" * 64


def _invalid(replay, replay_result="NOT_REQUESTED", through_seq=0):
    return {
        "integrity": "INVALID",
        "replay": replay_result,
        "through_seq": through_seq,
        "anchored": False,
    }


def _audit_key_at(root: dict, seq: int):
    for k in root["audit_keys"]:
        if k["from_seq"] <= seq and (k["through_seq"] is None or seq <= k["through_seq"]):
            return k
    return None


def _entry_valid(entry: dict, root: dict) -> bool:
    try:
        v_audit_entry(entry)
    except Exception:
        return False
    key = _audit_key_at(root, entry["body"]["seq"])
    if key is None or key["key_id"] != entry["key_id"]:
        return False
    if D("audit", entry["body"]) != entry["hash"]:
        return False
    try:
        return ed25519_verify(
            hex_decode(key["public_key"]), b64url_decode(entry["signature"]),
            S("audit", entry["body"]),
        )
    except Exception:
        return False


def _checkpoint_valid(cp: dict, root: dict) -> bool:
    try:
        v_checkpoint(cp)
    except Exception:
        return False
    key = _audit_key_at(root, cp["body"]["through_seq"] + 1) or _audit_key_at(
        root, cp["body"]["through_seq"]
    )
    if key is None or key["key_id"] != cp["key_id"]:
        return False
    if cp["body"]["tenant_id"] != root["tenant_id"] or cp["body"]["log_id"] != root["log_id"]:
        return False
    try:
        return ed25519_verify(
            hex_decode(key["public_key"]), b64url_decode(cp["signature"]),
            S("checkpoint", cp["body"]),
        )
    except Exception:
        return False


def _pin_key(p: dict) -> str:
    return canonical_string(p)


def verify_evidence(evidence: dict, root: dict, trusted_checkpoint, replay: bool) -> dict:
    through_seq = 0
    anchored = False

    try:
        v_evidence(evidence)
    except Exception:
        return _invalid(replay)

    # 1. supplied root must equal the caller-pinned root
    if not canonical_equal(evidence["root"], root):
        return _invalid(replay)

    # 2. bundle authority: bundles must verify as a chain under the root
    bundles = sorted(evidence["bundles"], key=lambda b: b["charter"]["version"])
    pins: dict = {}
    try:
        for i, b in enumerate(bundles):
            verify_bundle(b, root, bundles[:i])
            pin = {
                "charter_id": b["charter"]["charter_id"],
                "version": b["charter"]["version"],
                "charter_hash": D("charter", b["charter"]),
                "manifest_hash": D("manifest", b["manifest"]),
                "engine": b["charter"]["engine"],
            }
            pins[_pin_key(pin)] = {"charter": b["charter"], "manifest": b["manifest"]}
    except Exception:
        return {
            "integrity": "INVALID",
            "replay": "CONTEXT_MISSING" if replay else "NOT_REQUESTED",
            "through_seq": 0,
            "anchored": False,
        }

    # 3. start checkpoint
    prev_hash = ZERO_HASH
    expect_seq = 1
    if evidence["start"] is not None:
        if not _checkpoint_valid(evidence["start"], root):
            return {
                "integrity": "INVALID",
                "replay": "CONTEXT_MISSING" if replay else "NOT_REQUESTED",
                "through_seq": 0,
                "anchored": False,
            }
        prev_hash = evidence["start"]["body"]["head_hash"]
        expect_seq = evidence["start"]["body"]["through_seq"] + 1

    # 4. entries: schema → seq → hash → signature → linkage, in order
    integrity = "VALID"
    verified = []
    for e in evidence["entries"]:
        if not _entry_valid(e, root):
            integrity = "INVALID"
            break
        if e["body"]["seq"] != expect_seq or e["body"]["prev_hash"] != prev_hash:
            integrity = "INVALID"
            break
        verified.append(e)
        through_seq = e["body"]["seq"]
        prev_hash = e["hash"]
        expect_seq = e["body"]["seq"] + 1

    # 5. end checkpoint must be valid and name the verified head
    end_ok = False
    if integrity == "VALID":
        if not _checkpoint_valid(evidence["end"], root):
            integrity = "INVALID"
        elif (
            evidence["end"]["body"]["through_seq"] != through_seq
            or evidence["end"]["body"]["head_hash"] != prev_hash
        ):
            integrity = "INCOMPLETE"
        else:
            end_ok = True
    if end_ok and trusted_checkpoint is not None:
        anchored = canonical_equal(evidence["end"], trusted_checkpoint)

    # 6. optional replay
    replay_result = "NOT_REQUESTED"
    if replay and integrity == "VALID":
        replay_result = _replay_decisions(evidence, verified, pins)
    elif replay:
        replay_result = "CONTEXT_MISSING"

    return {
        "integrity": integrity,
        "replay": replay_result,
        "through_seq": through_seq,
        "anchored": anchored,
    }


def _replay_decisions(evidence: dict, entries: list, pins: dict) -> str:
    inputs = {i["request"]["request_id"]: i for i in evidence["inputs"]}
    pin_updates = {}
    for u in evidence["pin_updates"]:
        try:
            v_pin_update(u["update"])
            pin_updates[u["update"]["expected_revision"]] = u
        except Exception:
            return "CONTEXT_MISSING"
    current_pin = None
    for e in entries:
        ev = e["body"]["event"]
        if ev["type"] == "PinActivated":
            current_pin = ev["value"]["pin"]
            continue
        if ev["type"] not in ("CheckEvaluated", "CallDenied", "CallDispatched"):
            continue
        inp = inputs.get(ev["value"]["request_id"])
        if inp is None:
            return "INPUTS_MISSING"
        input_hash = D("input", {"request": inp["request"], "principal": inp["principal"]})
        if input_hash != ev["value"]["input_hash"]:
            return "MISMATCH"
        pin = e["body"]["policy_pin"] or current_pin
        if pin is None:
            return "CONTEXT_MISSING"
        ctx = pins.get(_pin_key(pin))
        if ctx is None:
            return "CONTEXT_MISSING"
        if (
            D("charter", ctx["charter"]) != pin["charter_hash"]
            or D("manifest", ctx["manifest"]) != pin["manifest_hash"]
        ):
            return "CONTEXT_MISSING"
        decision = evaluate({
            "charter": ctx["charter"], "manifest": ctx["manifest"],
            "active_pin": pin, "request": inp["request"],
            "principal": inp["principal"], "now": ev["value"]["evaluated_at"],
        })
        if not canonical_equal(decision, ev["value"]["decision"]):
            return "MISMATCH"
    return "MATCH"

File: src/evidence.ts
/**
 * Offline evidence verification (spec §4.4, §8.2, §13.2).
 * Validates supplied root equality to the caller-pinned root, entry
 * schema/seq/hash/signature, predecessor linkage, bundle authority, and the
 * final checkpoint — in that order. Never contacts the registry; never
 * dispatches on replay.
 */
import { canonicalString, type Json } from "./canon.js";
import { D, hexDecode, ed25519Verify, S, b64urlDecode } from "./crypto.js";
import type {
  AuditEntry, Bundle, Charter, Checkpoint, Decision, Evidence, Pin,
  RootFile, SignedPinUpdate, Verification,
} from "./types.js";
import { vEvidence, vAuditEntry, vCheckpoint, vPinUpdate } from "./schema.js";
import { evaluate } from "./evaluator.js";
import { canonicalEqual, verifyBundle } from "./bundle.js";

const ZERO_HASH = "0".repeat(64);

interface Fail {
  integrity: "INVALID" | "INCOMPLETE";
  throughSeq: number;
}

function auditKeyAt(root: RootFile, seq: number): { key_id: string; public_key: string } | null {
  return (
    root.audit_keys.find(
      (k) => k.from_seq <= seq && (k.through_seq === null || seq <= k.through_seq),
    ) ?? null
  );
}

function entryValid(entry: AuditEntry, root: RootFile): boolean {
  try {
    vAuditEntry(entry);
  } catch {
    return false;
  }
  const key = auditKeyAt(root, entry.body.seq);
  if (!key || key.key_id !== entry.key_id) return false;
  if (D("audit", entry.body as unknown as Json) !== entry.hash) return false;
  try {
    return ed25519Verify(
      hexDecode(key.public_key), b64urlDecode(entry.signature),
      S("audit", entry.body as unknown as Json),
    );
  } catch {
    return false;
  }
}

function checkpointValid(cp: Checkpoint, root: RootFile): boolean {
  try {
    vCheckpoint(cp);
  } catch {
    return false;
  }
  const key = auditKeyAt(root, cp.body.through_seq + 1) ?? auditKeyAt(root, cp.body.through_seq);
  if (!key || key.key_id !== cp.key_id) return false;
  if (cp.body.tenant_id !== root.tenant_id || cp.body.log_id !== root.log_id) return false;
  try {
    return ed25519Verify(
      hexDecode(key.public_key), b64urlDecode(cp.signature),
      S("checkpoint", cp.body as unknown as Json),
    );
  } catch {
    return false;
  }
}

function pinKey(p: Pin): string {
  return canonicalString(p as unknown as Json);
}

/**
 * verifyEvidence(evidence, root, trustedCheckpoint, replay).
 * `trustedCheckpoint` anchors the verified head; replay re-evaluates decision
 * events against caller-supplied private inputs.
 */
export function verifyEvidence(
  evidence: Evidence,
  root: RootFile,
  trustedCheckpoint: Checkpoint | null,
  replay: boolean,
): Verification {
  let integrity: Verification["integrity"] = "VALID";
  let throughSeq = 0;
  let anchored = false;

  try {
    vEvidence(evidence);
  } catch {
    return { integrity: "INVALID", replay: replay ? "NOT_REQUESTED" : "NOT_REQUESTED", through_seq: 0, anchored: false };
  }

  // 1. supplied root must equal the caller-pinned root
  if (!canonicalEqual(evidence.root, root)) {
    return { integrity: "INVALID", replay: "NOT_REQUESTED", through_seq: 0, anchored: false };
  }

  // 2. bundle authority: bundles must verify as a chain under the root
  const bundles = [...evidence.bundles].sort((a, b) => a.charter.version - b.charter.version);
  const pins = new Map<string, { charter: Charter; manifest: Bundle["manifest"] }>();
  try {
    for (let i = 0; i < bundles.length; i++) {
      verifyBundle(bundles[i]!, root, bundles.slice(0, i));
      const b = bundles[i]!;
      const pin: Pin = {
        charter_id: b.charter.charter_id, version: b.charter.version,
        charter_hash: D("charter", b.charter), manifest_hash: D("manifest", b.manifest),
        engine: b.charter.engine,
      };
      pins.set(pinKey(pin), { charter: b.charter, manifest: b.manifest });
    }
  } catch {
    return { integrity: "INVALID", replay: replay ? "CONTEXT_MISSING" : "NOT_REQUESTED", through_seq: 0, anchored: false };
  }

  // 3. start checkpoint (if present) must verify and name the entry prefix base
  let prevHash = ZERO_HASH;
  let expectSeq = 1;
  if (evidence.start !== null) {
    if (!checkpointValid(evidence.start, root)) {
      return { integrity: "INVALID", replay: replay ? "CONTEXT_MISSING" : "NOT_REQUESTED", through_seq: 0, anchored: false };
    }
    prevHash = evidence.start.body.head_hash;
    expectSeq = evidence.start.body.through_seq + 1;
  }

  // 4. entries: schema → seq → hash → signature → linkage, in order
  const verified: AuditEntry[] = [];
  for (const e of evidence.entries) {
    if (!entryValid(e, root)) {
      integrity = "INVALID";
      break;
    }
    if (e.body.seq !== expectSeq || e.body.prev_hash !== prevHash) {
      integrity = "INVALID";
      break;
    }
    verified.push(e);
    throughSeq = e.body.seq;
    prevHash = e.hash;
    expectSeq = e.body.seq + 1;
  }

  // 5. end checkpoint must be valid and name the verified head
  let endOk = false;
  if (integrity === "VALID") {
    if (!checkpointValid(evidence.end, root)) {
      integrity = "INVALID";
    } else if (
      evidence.end.body.through_seq !== throughSeq ||
      evidence.end.body.head_hash !== prevHash
    ) {
      // bytes absent: supplied evidence does not reach the signed bound
      integrity = "INCOMPLETE";
    } else {
      endOk = true;
    }
  }
  if (endOk && trustedCheckpoint !== null) {
    anchored = canonicalEqual(evidence.end, trustedCheckpoint);
  }

  // 6. optional replay: match inputs by request_id, re-evaluate under recorded pin/time
  let replayResult: Verification["replay"] = "NOT_REQUESTED";
  if (replay && integrity === "VALID") {
    replayResult = replayDecisions(evidence, verified, pins);
  } else if (replay) {
    replayResult = integrity === "INCOMPLETE" ? "CONTEXT_MISSING" : "CONTEXT_MISSING";
  }

  return { integrity, replay: replayResult, through_seq: throughSeq, anchored };
}

function replayDecisions(
  evidence: Evidence,
  entries: AuditEntry[],
  pins: Map<string, { charter: Charter; manifest: Bundle["manifest"] }>,
): Verification["replay"] {
  const inputs = new Map(evidence.inputs.map((i) => [i.request.request_id, i]));
  const pinUpdates = new Map<number, SignedPinUpdate>();
  for (const u of evidence.pin_updates) {
    try {
      vPinUpdate(u.update);
      pinUpdates.set(u.update.expected_revision, u);
    } catch {
      return "CONTEXT_MISSING";
    }
  }
  // pin timeline: PinActivated events give revision→pin
  const activatedPins = new Map<number, Pin>();
  let currentPin: Pin | null = null;
  for (const e of entries) {
    if (e.body.event.type === "PinActivated") {
      currentPin = e.body.event.value.pin;
      activatedPins.set(e.body.event.value.revision, currentPin);
    }
  }
  currentPin = null;
  for (const e of entries) {
    if (e.body.event.type === "PinActivated") {
      currentPin = e.body.event.value.pin;
      continue;
    }
    const t = e.body.event.type;
    if (t !== "CheckEvaluated" && t !== "CallDenied" && t !== "CallDispatched") continue;
    const ev = e.body.event.value;
    const input = inputs.get(ev.request_id);
    if (!input) return "INPUTS_MISSING";
    const inputHash = D("input", { request: input.request, principal: input.principal } as unknown as Json);
    if (inputHash !== ev.input_hash) return "MISMATCH";
    const pin = e.body.policy_pin ?? currentPin;
    if (!pin) return "CONTEXT_MISSING";
    const ctx = pins.get(pinKey(pin));
    if (!ctx) return "CONTEXT_MISSING";
    if (D("charter", ctx.charter) !== pin.charter_hash || D("manifest", ctx.manifest) !== pin.manifest_hash) {
      return "CONTEXT_MISSING";
    }
    const decision: Decision = evaluate({
      charter: ctx.charter, manifest: ctx.manifest, active_pin: pin,
      request: input.request, principal: input.principal, now: ev.evaluated_at,
    });
    if (!canonicalEqual(decision, ev.decision)) return "MISMATCH";
  }
  return "MATCH";
}
